flink-runtime/src/main/java/org/apache/flink/streaming/api/functions/source/InputFormatSource.java

```diff
@@ -0,0 +1,314 @@
+            SplitEnumeratorContext<SourceSplit> context, Void checkpoint) throws Exception {
+        return new InputFormatSplitEnumerator<>(format, context);
+    }
+
+    @Override
+    public SimpleVersionedSerializer<SourceSplit> getSplitSerializer() {
+        return new SimpleVersionedSerializer<>() {
+            @Override
+            public int getVersion() {
+                return 0;
+            }
+
+            @Override
+            public byte[] serialize(SourceSplit split) throws IOException {
+                return InstantiationUtil.serializeObject(split);
+            }
+
+            @Override
+            public SourceSplit deserialize(int version, byte[] serialized) throws IOException {
+                try {
+                    return InstantiationUtil.deserializeObject(
+                            serialized, SourceSplit.class.getClassLoader());
+                } catch (ClassNotFoundException e) {
+                    throw new IOException("Failed to deserialize SourceSplit.");
+                }
+            }
+        };
+    }
+
+    @Override
+    public SimpleVersionedSerializer<Void> getEnumeratorCheckpointSerializer() {
+        return new SimpleVersionedSerializer<>() {
+            @Override
+            public int getVersion() {
+                return 0;
+            }
+
+            @Override
+            public byte[] serialize(Void obj) {
+                return new byte[0];
+            }
+
+            @Override
+            public Void deserialize(int version, byte[] serialized) {
+                return null;
+            }
+        };
+    }
+
+    @Override
+    public SourceReader<OUT, SourceSplit> createReader(SourceReaderContext readerContext)
+            throws Exception {
+        Preconditions.checkState(
+                readerContext instanceof RichSourceReaderContext,
+                "Source reader context must be an instance of RichSourceReaderContext");
+        RichSourceReaderContext richSourceReaderContext = (RichSourceReaderContext) readerContext;
+        RuntimeContext runtimeContext = richSourceReaderContext.getRuntimeContext();
+        Counter completedSplitsCounter =
+                runtimeContext.getMetricGroup().counter("numSplitsProcessed");
+
+        return new SourceReader<>() {
+            private Queue<SourceSplit> remainingSplits;
+            private boolean noMoreSplits;
+            private boolean isFormatOpen;
+            private OUT lastElement;
+
+            @Override
+            public void start() {
+                this.remainingSplits = new LinkedList<>();
+                this.noMoreSplits = false;
+                this.isFormatOpen = false;
+                this.lastElement = null;
+
+                if (format instanceof RichInputFormat) {
+                    ((RichInputFormat<?, ?>) format).setRuntimeContext(runtimeContext);
+                }
+                format.configure(new Configuration());
```

Comment:
This doesn't look right, can we get the actual config from the source or runtime context?

Reply:
This is coming from the old implementation so pretty sure that it works. Let me check whether we can enhance it now.

Reply:
Enhanced.